Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "RenderTreeUpdaterViewTransition.h"
+
+#include "RenderDescendantIterator.h"
+#include "RenderElement.h"
+#include "RenderStyleInlines.h"
+#include "RenderTreeUpdater.h"
+#include "RenderView.h"
+#include "StyleTreeResolver.h"
+#include "ViewTransition.h"
+
+namespace WebCore {
+
+RenderTreeUpdater::ViewTransition::ViewTransition(RenderTreeUpdater& updater)
+    : m_updater(updater)
+{
+}
+
+// The contents and ordering of the named elements map should remain stable during the duration of the transition.
+// We should only need to handle changes in the `display` CSS property by recreating / deleting renderers as needed.
+void RenderTreeUpdater::ViewTransition::updatePseudoElementTree(RenderElement& documentElementRenderer)
```

Comment:
Is there a good test coverage for mutations? You should ensure these pseudo elements are added to the fuzzers soon.

Reply:
WPTs cover mutations to some extent, but right now proper rendering test coverage can't be done until we hook this up to animation timing because pseudo-elements won't appear for long enough (this is what I'm planning to look into next).